RFC: IRI v1.8.1 transactions and bundles (#20)
iotaledger/bee-rfcs: text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

# Summary

The fundamental communication unit in the IOTA protocol is the transaction. Everything, including payment settlements
and plain data, is propagated through the IOTA network in transactions.

A transaction is `8019` trits and the payload - `sig_or_msg` field - is `6561` trits. This payload can hold a signature
fragment or a message fragment. Since it has a limited size, a user often needs more than one transaction to fulfil
their operation, for example signatures with security level `2` or `3` don't fit in a single transaction and
user-provided message may exceed the allowance so they need to be fragmented across multiple transactions. Moreover,
input/output transactions doesn't make sense on their own because they would change the total amount of the ledger so
they have to be paired with other input/output transactions that together will balance the total value to zero.

For these reasons, transactions have to be processed as a whole, in groups called bundles. A bundle is an atomic
operation in the sense that either all or none of its transactions are accepted by the network. Even single
transactions are propagated through the network within a bundle.

> By analogy with TCP, a bundle corresponds to a stream, and a transaction corresponds to a packet.

Reviewer: Are we expecting RFC readers to be knowledgeable in TCP to understand this analogy? Apart from that the analogy is bad because a bundle is not a stream (which insinuates potential endlessness), it's much more similar to a container.

+ [What is a bundle?](https://docs.iota.org/docs/getting-started/0.1/introduction/what-is-a-bundle)
+ [Bundles and transactions](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+ [Structure of a transaction](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-transaction)
+ [Structure of a bundle](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-bundle)

# Motivation

## Transaction

IOTA is a transaction settlement and data transfer layer for IoT (the Internet of Things). Messages on the network are
`Bundle`s of individual `Transaction`s, which in turn are sent one at a time and are stored in a distributed ledger
called the *Tangle*. Each `Transaction` encodes data such as sender and receiver addresses, referenced transactions in
the Tangle, `Bundle` hash, timestamps, and other information required to verify and process each transaction. With the
transaction being the fundamental unit that is sent over the network, we need to represent it in memory.

At the time of this RFC, the transaction format used by the IOTA network is defined by [release v1.8.1 of the IOTA

> Reference Implementation, IRI](`iri v1.8.1`), commit `e1776fbad5d90df86a26402f9025e4b0b2ef7d3e`. The transaction format

Reviewer 2: ```suggestion
Reference Implementation, IRI](`iri v1.8.1`), commit [e1776fbad5d90df86a26402f9025e4b0b2ef7d3e](https://github.com/iotaledger/iri/tree/e1776fbad5d90df86a26402f9025e4b0b2ef7d3e). The transaction format
```

Author: Not sure what commit you're reviewing but I simply removed mention of the commit hash. It was not necessary since already included in the IRI version.

Reviewer 2: It's the same hash but with link on doc. If you want to put the commit hash here, better have a link instead of text only.

Author: I totally removed the commit hash.

Author: Ahhhh sorry I didn't see the line! There was two place with the exact same hash and I removed the other! Let me do that right now.

access to fields is permitted.

## Bundle

<!-- TODO -->

# Detailed design

## Transaction

A transaction is a sequence of 15 fields of constant length. A transaction has a total length of 8019 trits. The fields'
name, description, and size (in trits) is summarized in the table below, in their order of appearance in a transaction.

| Name                              | Description                                            | Size (in trits) |
| ---                               | ---                                                    | ---             |
| `signature_or_message_fragment`   | contains the signature of the transfer                 |                 |
|                                   | or user-defined message data                           | 6561            |
| `address`                         | receiver (output) if value > 0,
|                                   | or sender (input) if value < 0                         | 243             |
| `value`                           | the transferred amount in IOTA                         | 81              |

> | `obsolete_tag`                    | another arbitrary user-defined tag                     | 81              |

Reviewer 2: don't think `obsolete_tag` is defined by the user, it's related to the M bug and will be changed in finalization.

Author: I rephrased, thanks

| `timestamp`                       | the time when the transaction was issued               | 27              |
| `current_index`                   | the position of the transaction in its bundle          | 27              |
| `last_index`                      | the index of the last transaction in the bundle        | 27              |
| `bundle`                          | the hash of the bundle essence                         | 243             |
| `trunk`                           | the hash of the first transaction referenced/approved  | 243             |
| `branch`                          | the hash of the second transaction referenced/approved | 243             |

> | `tag`                             | arbitrary user-defined value                           | 81              |

Reviewer 2: ```suggestion
| `tag`                             | arbitrary user-defined tag                           | 81              |
```

Author: Janis was using `value` as content, not as field name, I think it's fine.

| `attachment_timestamp`            | the timestamp for when Proof-of-Work is completed      | 27              |
| `attachment_timestamp_lowerbound` | *not specified*                                        | 27              |
| `attachment_timestamp_upperbound` | *not specified*                                        | 27              |
| `nonce`                           | the Proof-of-Work nonce of the transaction             | 81              |

Each transaction is uniquely identified by its *transaction hash*, which is calculated based on all fields of the

> transaction. Note that the transaction hash is not part of the transaction.

Reviewer: Just for the sake of discussion:
Conceptually the `Nonce` shouldn't be part of the transaction either, because it's part of a spam protection mechanism. 
Again, just a point to talk about for future design considerations.

Author: I get your point. But this RFC is about IRI 1.8.1 transaction. If nonce is not part of the transaction then you somehow have to consider global identities and opens up Sybil attacks. I believe this is atm a research topic.

+ each transaction has a `current_index` and a `last_index` and `current_index` goes from `0` to `last_index`, a bundle
can then simply be represented by a data structure that contiguously keeps the order like `Vec`;
+ each transaction is chained to the next one through its `trunk` which means we can consider data structures like
`HashMap` or `BTreeMap`;

For this reason, we hide this as an implementation detail and instead provide a newtype:

```rust
struct Transactions(Vec<Transaction>)
//struct Transactions(HashMap<Transaction>)
//struct Transactions(BTreeMap<Transaction>)
```

Then the `Bundle` type looks like:

```rust
struct Bundle {
    transactions: Transactions
}
```

And its implementation should only allow to retrieve transactions:

```rust
impl Bundle {    
    pub fn transactions(&self) -> &Transactions {
        &self.transactions
    }
}
```

### BundleBuilder

The `BundleBuilder` offers a simple and convenient way to build bundles:

```rust
pub struct BundleBuilder {
    transactions: Vec<TransactionBuilder>
}
```

```rust
impl BundleBuilder {    
    pub fn calculate_hash(&self) {
      unimplemented!()
    }

>     pub fn finalise(&self) {

Reviewer 2: for consistency
```suggestion
    pub fn finalize(&self) {
```

Author: Will do

`M` or `13` because it could expose a significant part of the private key, weakening the signature. The bundle hash is
then repetitively generated with a slight modification until its normalisation doesn't contain a `M`.*

#### Sign

*Client side operation.*

Signing a bundle allow you to prove that you are the owner of the address you are trying to move funds from. With no
signature or a bad signature, the bundle won't be considered valid and the funds won't be moved. Only the owner of the
right seed is able to generate the right signature for this address.

Pseudocode:

```
sign(bundle, seed, inputs)
| current_index = 0
|
| for transaction in bundle
| | if transaction.value < 0
| | | if transaction.current_index >= current_index
| | | | input = inputs[transaction.address]
| | | | fragments = sign(seed, input.index, input.security, transaction.bundle)
| | | | for fragment in fragments
| | | | | bundle.at(current_index).signature = fragment
| | | | | current_index = current_index + 1
| | else
| | | current_index = current_index + 1
```

*Since signature size depends on the security level, a single signature can spread out to up to 3 transactions.

> `inputs` is an object that contain all unused addresses of a seed with a sufficient balance.*

Reviewer 2: ```suggestion
`inputs` is an object that contains all unused addresses of a seed with a sufficient balance.*
```

Author: Thanks

#### Proof of Work

*Client side operation.*

Proof of Work (PoW) allows your transactions to be accepted by the network. On the IOTA network, PoW is only a rate

> control mechanism. Doing PoW on a bundle means doing PoW on each of its transactions and setting trunks and branch

Reviewer 2: ```suggestion
PoW is meant to prevent Sybil attack. Doing PoW on a bundle means doing PoW on each of its transactions and setting trunks and branch
```

Author: PoW is a rate control mechanism, not an anti-Sybil protection. Mana is.

+ Feature name: `iri_v181_transaction_and_bundle`
+ Start date: 2019-09-06
+ RFC PR: [iotaledger/bee-rfcs#3](https://github.com/iotaledger/bee-rfcs/pull/3),
  [iotaledger/bee-rfcs#18](https://github.com/iotaledger/bee-rfcs/pull/18),
  [iotaledger/bee-rfcs#20](https://github.com/iotaledger/bee-rfcs/pull/20)
+ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)

# Summary

The fundamental communication unit in the IOTA protocol is the transaction. Messages, including payment settlements and
plain data, are propagated through the IOTA network in transactions. Message that are too large are split up into bundles

> of several transactions and sent one by one.

Reviewer: I have a proposal for that paragraph which I find more clear: 

Messages in the IOTA network can either represent payment settlements or serve as plain data carriers. In IOTA terminology messages are referred to as 'Bundles', which - depending on the message size - consist of one or more transactions. A transaction is the smallest communication unit. It has a fixed size, while messages in general can have arbitrary size. (maybe support this explanation with an example, e.g. security 3 transfer)

Reviewer 3: Agree Alex, sounds better and is easier to understand.

This RFC proposes a `Transaction` type and a `Bundle` type to represent the transaction and bundle formats used by the

> IOTA Reference Implementation as of release [`iri v1.8.1`]. We also propose the `TransactionBuilder`,

Reviewer: * I think for consistency it's better to always stick to 3rd person ("this RFC proposes ...", "it also proposes", ...) rather than 1st person plural ("we propose ...") 
* Message[s] that are too large ...

Reviewer 4: Fixed for all instances of *we* in the text.

`IncomingBundleBuilder`, and `OutgoingBundleBuilder` types following the builder pattern to construct both message
types, respectively. We distinguish between incoming and outgoing bundles because one is concerned with verifying an
existing message being received, while the other one is concerned with constructing a new message intended to be sent,

> which includes setting a number of fields like the bundle hash.

Reviewer: I think fields are set in both cases. The differentiating factor is, that for the outgoing bundle we actually have to create/calculate those fields for the first time while for the incoming bundle those fields can just be read from the wire bytes.

Reviewer 4: Reformulated that section.

Useful links:

+ [Trinary](https://docs.iota.org/docs/dev-essentials/0.1/concepts/trinary)
+ [What is a transaction?](https://docs.iota.org/docs/getting-started/0.1/introduction/what-is-a-transaction)
+ [What is a bundle?](https://docs.iota.org/docs/getting-started/0.1/introduction/what-is-a-bundle)
+ [Bundles and transactions](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+ [Structure of a transaction](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-transaction)
+ [Structure of a bundle](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-bundle)
+ [Security levels](https://docs.iota.org/docs/dev-essentials/0.1/references/security-levels)

# Motivation

> IOTA is a transaction settlement and data transfer layer for the Internet of Things. Messages on the network are

Reviewer: we should not restrict ourselves to IoT. IOTA should also be for the classic Internet.

Reviewer 5: I think the term IOT already includes classic internet and has much boarder meaning.

Author: We usually differentiate IoT and Internet though, I agree with @Alex6323, I'm not sure we should emphasise on IoT here.

Reviewer 5: In general, IoT consists of end devices (sensors and actuators), gateways, and the clouds (including internet). The applications of an IoT system usually include sensing, data analytics, decision making, and actions. So regarding IOTA as a data transfer layer for IoT makes sense.

Reviewer 4: To me, the term IoT also includes the general internet. At the end of the day, in IoT you are also sending information from full sized computers to embedded devices and the other way round.

in another would be constructed like this:

```rust
let mut outgoing_bundle_builder = OutgoingBundleBuilder::new();

let mut withdrawing_tx_builder = TransactionBuilder::new();
let mut depositing_tx_builder = TransactionBuilder::new();

// The transactions here only need to set addresses, values, and tags. All other fields, such as the signature
// of the withdrawing transaction, will be filled in by the bundle builder.
withdrawing_tx_builder
    .address(address_providing_funds)
    .value(value_to_transfer)
    .tag(custom_tag);

depositing_tx_builder
    .address(address_receiving_funds)
    .value(value_to_transfer)
    .tag(custom_tag);

let outgoing_bundle = outgoing_bundle_builder
    .push(withdrawing_tx_builder)
    .push(depositing_tx_builder)
    // Seal the bundle, adding indices to its transactions and assigning a bundle hash. This also prevents
    // pushing more transactions into the bundle.
    .seal()?
    // Withdrawal transactions are signed, providing some signature scheme, seed, and wallet. If none are present, this
    // step can be skipped.

>     .sign(signature_scheme, seed, wallet)?

Reviewer 3: Does this suggest that multiple signature schemes are allowed like WOTS/ECC/...
If so, is it really needed?

Reviewer 4: That depends on how we implement `sign`. If `sign` takes a `S: SignatureScheme`, then yeah --- it would allow for different schemes. If it takes a `WOTS`, then no. Either way, we'd probably pass the scheme in explicitly because it might need to be configured.

But let's see what its RFC brings. I have made a note in the RFC.

[`iri v1.8.1`]: https://github.com/iotaledger/iri/releases/tag/v1.8.1-RELEASE

# Detailed design

This RFC proposes the implementation of the following types:

+ `Transaction` to represent the transaction format as of `iri v1.8.1`;
+ `TransactionBuilder`, a builder pattern to create a `Transaction`;
+ `Bundle` to represent a set of `Transaction`s;
+ `IncomingBundleBuilder`, a builder pattern to create `Bundle` from a set of already created `Transaction`s, such as
  those coming in over the wire;
+ `OutgoingBundleBuilder` and `SealedBundleBuilder`, two builder patterns working in tandem to create a `Bundle` from
  `TransactionBuilder`s, i.e. transactions that are not yet fully constructed and will be finalized during the
  construction process of the entire `Bundle`.

In this proposal we keep the representation of the various types relatively simple and flat. We assume that the user is
responsible for pushing the appropriate number of transaction drafts into a bundle builder and don't encode this on the
type level. During the construction phase we check that all transactions are consistent and adhere to IOTA convention,
but we don't, for example:

> + attempt to represent different the different types of transactions on the type level;

Reviewer: remove one 'different'

fragmented across multiple transactions. Moreover, because the total amount of tokens stored in the ledger has to stay
constant, *input transactions* (which are called thus because an address is *put into* a transfer as a source of tokens,
thus removing them from the address) have to always be matched with *output transactions* such that their total value is
zero. For these reasons, transactions have to be processed as a whole in groups called bundles. A bundle is an atomic
operation in the sense that either all or none of its transactions are accepted by the network. Even single transactions
are propagated through the network within a bundle.

The `Transaction` and `Bundle` types are intended to be final and immutable. This contract is enforced by not exposing
any methods that allow manipulating their fields directly. `Transaction`s are intended to be constructed through
the `TransactionBuilder` builder pattern, while `Bundle`s can be created either via the `IncomingBundleBuilder` or
`OutgoingBundleBuilder`. The `IncomingBundleBuilder` takes in existing `Transaction`s, compiles them, and verifies

> their veracity before constructing a `Bundle` type. The `OutgoingBundleBuilder` takes in `TransactionBuilder`s,

Reviewer: i'm not sure 'veracity' is the correct term here. Sounds more like a metaphysical term. What about 'validity'?

Reviewer 4: Fixed. That formulation was awful. :)

and is primarily responsible for filling in all information that can only be set in the actual finalization
of the outgoing message into a `Bundle`, such as the bundle hash.

## Transaction

A transaction is a sequence of 15 fields of constant length. A transaction has a total length of 8019 trits. The fields'
name, description, and size are summarized in the table below, in their order of appearance in a transaction.

> | Name                              | Description                                             | Size (in trits) | Set by (functions detailed below)             |

Author: Need to change the functions name in `Set by` column. And remove `add_message`, `add_input` and `add_output`.

Reviewer 4: I have removed the column entirely. It does not contain a lot of information at the moment, and I see this to be more relevant when describing the actual implementation.

| `nonce`                           | the Proof-of-Work nonce of the transaction              | 81              | `calculate_proof_of_work`                     |

Each transaction is uniquely identified by its *transaction hash*, which is calculated based on all fields of the
transaction. Note that the transaction hash is not part of the transaction.

Because modifying the fields of a transaction would invalidate its hash, `Transaction` is immutable after creation.

### `Transaction` struct

A transaction is represented by the `Transaction` struct and follows the structure presented in the table above:

```rust
pub struct Transaction {
    signature_or_message_fragment: SignatureOrMessageFragment,
    address: Address,
    value: Value,
    obsolete_tag: Tag,
    timestamp: Timestamp,
    current_index: Index,
    last_index: Index,
    bundle: BundleHash,
    trunk: TransactionHash,
    branch: TransactionHash,
    tag: Tag,
    attachment_timestamp: Timestamp,
    attachment_timestamp_lower_bound: Timestamp,
    attachment_timestamp_upper_bound: Timestamp,
    nonce: Nonce
}
```

> The `Transaction` type only contains two constructor methods getter methods for read-only borrows of its fields.

Reviewer: - two constructor functions and some getter methods for read-only borrows of its fields

Reviewer 4: Fixed and removed one getter method. The implementation phase should define more constructors.

```rust
impl Transaction {
    /// Create a `Transaction` from a reader.
    pub fn from_reader<R: std::io::Read>(reader: R) -> Result<Self, TransactionError> {
        unimplemented!()
    }

    /// Create a `Transaction` from a slice of bytes
    pub fn from_slice(stream: &[u8]) -> Result<Self, TransactionError> {
        unimplemented!()
    }

    pub fn signature_or_message_fragment(&self) -> &SignatureOrMessageFragment {
        &self.signature_fragments
    }

    pub fn address(&self) -> &Address {
        &self.address
    }

    pub fn value(&self) -> &Value {
        &self.value
    }
}
```

We treat the types of the `Transaction` struct's fields as opaque newtypes for now so that we can flesh them out during
the implementation phase or in future RFCs. Because this RFC implements the transaction format as of iri v1.8.1, the

> newtypes shall be constructed from byte slices of a certain length matching that of the reference implementation.

Reviewer: a Java byte is an 'i8' in Rust, not an 'u8'. So if we want to match the reference implementation we should use 'i8' in this RFC as the proper equivalent. It also simplifies calculations as we tell the compiler about how we expect the bits in the byte to be interpreted (0xFF as -1, not as 255).

Reviewer 4: I have added a sentence about converting from `&[i8]`. In general, I'd like to keep the representation as bytes, but conversion form `&[i8]` is fine. Also, we can change this down the road if we realize that this is not nice.

        None
    }
}

impl Tag {
    /// Unchecked conversion from a byte slice.
    ///
    /// Panics if the slice has the wrong length, and does not check if the bytes correctly encode
    /// a trit.
    fn from_unchecked(bytes: &[u8]) -> Self {
        let mut array = [0u8; 81];
        // Panics if `bytes` is not of length 81
        Tag(array.copy_from_slice(bytes))
    }
}

impl<'_> TryFrom<&'_ [u8]> for Tag {
    type Error = TagError;

    fn try_from(value: &'_ [u8]) -> Result<Self, Self::Error> {
        if value.len() != 81 {
            Err(TagError::WrongLength { expected: 81, given: value.len()})?;
        }

        for byte in value {
            match byte {
                0b0000_0000 | 0b1111_1111 | 0b0000_0001 => {},
                _ => Err(TagError::InvalidBinaryEncodedTrit)?,
        }

        Ok(Tag::from_unchecked(value))
    }
}
```

### `TransactionBuilder` struct

The `TransactionBuilder` allows setting all the fields of a transaction, and verifies and builds a correct `Transaction`
type. The order of the fields in `TransactionBuilder` follows the table showing the different fields of a transaction,
the same as `Transaction`.

```rust

> pub struct TransactionBuilder {

Reviewer 3: Just a little thing here, but the TransactionBuilder actually doesn't build the Transaction, it's the BundleBuilder which fills indices, bundle hash, etc. and finally builds it right?
So maybe we should call it `TransctionDraft` instead of `TransactionBuilder`, as you suggested before
Letting the BundleBuilder then direclty operate on the `TransctionDraft` (with all fields there: bundlehash, indices, timestamps, ... ) seems clean to me :+1:

`trusted_bytes: &[u8]`, calling `builder.tag(Tag::from_unchecked(trusted_bytes))` sets the `tag` field to a value that
was not verified to only contained valid bytes.

```rust
enum TransactionBuilderError {
    Tag(TagError),
}

impl From<TagError> for Error {
    fn from(value: TagError) -> Self {
        Error::Tag(value)
    }
}

impl fmt::Display for TransactionBuilderError {
    fn display(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use TransactionBuilderError::*;

        let msg = match self {
            Tag(_) => "Failed setting tag from byte input.",
        };

        fmt!(f, &msg)
    }
}

impl error::Error for TransactionBuilderError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Tag(e) => Some(e),
        }
    }
}

impl TransactionBuilder {
    fn tag<T: TryInto<Tag>>(&mut self, tag: T) -> Result<&mut Self, TransactionBuilderError> {
        self.tag.replace(tag.try_into()?);
        self
    }

    /// Signs the `TransactionBuilder` by using the signature scheme `S` with some seed and input.
    ///
    /// TODO: This depends on one trait `SignatureScheme` which is not implemented yet.

>     fn sign<I, S>(&mut self, signature_scheme: S, seed: S::Seed, input: I) -> Result<&mut Self, TransactionBuilderError>

Author: It's not possible to sign individual transactions if you consider security level > 1

Reviewer 4: Good point, I have removed the function on `TransactionBuilder` and made a note of this in the open questions.

    /// The bundle hash is then repetitively generated with a slight modification until its normalization doesn't contain a `M`.
    /// This modification is usually operated by incrementing the `obsolete_tag` of the first transaction of the bundle since
    /// this field is not being used for any other reason. Useful links:
    /// [Why is the normalized hash considered insecure when containing the char 'M'](https://iota.stackexchange.com/questions/1241/why-is-the-normalized-hash-considered-insecure-when-containing-the-char-m).
    /// 
    /// *Since signature size depends on the security level, a single signature can spread out to up to 3 transactions.
    /// `inputs` is an object that contains all unused addresses of a seed with a sufficient balance.*
    /// 
    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
    ///       how the sponge is thought to be used.
    pub fn seal<S: Sponge<Output=BundleHash>>(self, sponge: S) -> Result<SealedOutgoingBundleBuilder, OutgoingBundleBuilderError> {
        unimplemented!()

        let mut current_index = 0;

>         for transaction_draft in bundle_builder {

Author: We use both `transaction_draft` and `transaction_builder` in this RFC, can we harmonize ?

Reviewer 4: Replaced in favour of `transaction_builder`.

    /// clear that these transactions should be processed as a whole.
    ///
    /// The hash of a bundle is derived from the bundle essence of each of its transactions. A bundle essence is a `486` trits
    /// subset of the transaction fields.
    ///
    /// | Name          | Size      |
    /// | ------------- | --------- |
    /// | address       | 243 trits |
    /// | value         | 81 trits  |
    /// | obsolete_tag  | 81 trits  |
    /// | timestamp     | 27 trits  |
    /// | current_index | 27 trits  |
    /// | last_index    | 27 trits  |
    ///
    /// The bundle hash is generated with a sponge by iterating through the bundle, from `0` to `last_index`, absorbing the
    /// bundle essence of each transaction and eventually squeezing the bundle hash from the sponge.
    ///
    /// NOTE: `iri v1.8.1` uses `Kerl` as a sponge.
    ///
    /// NOTE: This function is defined on: `SealedOutgoingBundleBuilder`, `OutgoingBundleBuilder`, and `IncomingBundleBuilder`.
    ///
    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
    ///       how the sponge is thought to be used.

>     pub fn calculate_hash<S: Sponge<Output=BundleHash>>(&self, mut sponge: S) -> S::Output {

Author: Is it worth thinking of a default implementation for this ? That's quite some duplication atm.

Reviewer 4: I am not totally sure about this, so I am going to leave this until after we have a `Sponge` trait.

I have made a not of this in the RFC.

    /// [`getTransactionsToApprove`]: https://docs.iota.org/docs/node-software/0.1/iri/references/api-reference#gettransactionstoapprove
    ///
    /// TODO: This function relies on some `Sponge` and `ProofOfWork` traits, which are not yet defined. The code below lays out
    ///       how proof of work is thought to be used.
    pub fn chain_and_attach<S: Sponge, P: ProofOfWork>(
        mut self,
        trunk_tip: TransactionHash,
        branch_tip: TransactionHash,
        proof_of_work: P,
        sponge: S,
    ) -> Result<Self, OutgoingBundleError> {
        unimplemented!()

        // Iterates over all transactions starting from the head, setting trunks, branches, timestamps, tags, and nonce.
        //
        // NOTE: We do not provide an implementation here.
        let builder_iterator = self.iter_from_head();

        // Set the head transaction
        let head_transaction = builder_iterator.next().ok_or("builder does not contain any transaction builders")?;
        head_transaction.trunk(trunk_tip);
        head_transaction.branch(branch_tip);

        head_transaction.attachment_timestamp = timestamp();
        head_transaction.attachment_timestamp_lower = 0;

>         head_transaction.attachment_timestamp_upper = 3812798742493;

Reviewer 5: It's good to add explanation for this number.

Reviewer 4: Added a `FIXME` to the code because I am not sure where that number comes from.

        if head_transaction.tag.is_empty() {
            head_transaction.tag = transaction_draft.obsolete_tag;
        }
        head_transaction.set_nonce_from_proof_of_work(proof_of_work);

        // `next_local_trunk` is the hash that will be used as value of the current transaction pointing to the previous
        // one.
        let mut next_local_trunk = head_transaction.calculate_hash(sponge);

        // Set the rest of the transactions
        for transaction_draft in builder_iterator {
            transaction_draft.branch(trunk_tip);
            transaction_draft.trunk(branch_tip);

            transaction_draft.attachment_timestamp = timestamp();
            transaction_draft.attachment_timestamp_lower = 0;
            transaction_draft.attachment_timestamp_upper = 3812798742493;

            if transaction_draft.tag.is_empty() {
                transaction_draft.tag = transaction_draft.obsolete_tag;
            }

            transaction_draft.set_nonce_from_proof_of_work(proof_of_work);

>             next_local_trunk = transaction_draft.calculate_hash(sponge);

Author: You are setting `next_local_trunk` but not using it.

Reviewer 4: Fixed.

    ///       how the signing scheme is thought to be used.
    ///
    /// TODO: This function relies on some `Wallet` trait, which is not yet defined. It's supposed to provide some
    ///       `f: Wallet W -> Tx Address A -> Signature Scheme Inputs I`
    pub fn sign<I, S, W>(self, signature_scheme: S, seed: S::Seed, wallet: Wallet) -> Result<Self, OutgoingBundleError>
    where
        S: SignatureScheme<I>,
        W: Wallet<A, Output=S::I>,
    {
        unimplemented!();

        // Pseudocode how `sign` should be used.

        let mut current_index = 0

        // TODO: Explain what this loop is trying to achieve.
        for transaction_draft in self {
            if transaction_draft.value < 0 {
                if transaction_draft.current_index >= current_index {
                    let input = inputs[transaction_draft.address];

                    // NOTE: The specific arguments to the signature scheme depend on the implementation of `SignatureScheme`.
                    //
                    // NOTE: The specific arguments to wallet depend on the implementation of `Wallet`.

>                     let fragments = transaction_draft.sign(signature_scheme, seed, wallet.get(transaction_draft.address));

Author: It's not on the `transaction_draft` scope. Will be on private key scope from Signing Scheme.

Reviewer 4: I have made a note of this in the sample code.

            if transaction_draft.bundle_hash != bundle_hash {
                Err(InvalidHash)?
            }

            if abs(transaction_draft.value) > IOTA_SUPPLY {
                Err(InvalidTransactionValue)?
            }

            value += transaction_draft.value;
            if abs(value) > IOTA_SUPPLY {
                Err(InvalidValue)?
            }

            if transaction_draft.current_index != current_index++ {
                Err(InvalidIndex)?
            }

            if transaction_draft.last_index != last_index {
                Err(InvalidIndex)?
            }

            if transaction_draft.value != 0 && transaction_draft.address.last != 0 {
                Err(InvalidAddress)?
        }

        if value != 0 {
            Err(InvalidValue)?
        }

        if bundle_hash != self.calculate_hash() {
            Err(InvalidHash)?
        }

        // NOTE: We do not provide an example of `validate_withdrawal_transactions`. Its purpose is to walk
        //       through those transactions in the bundle builder and verify their signatures.
        bundle_builder.validate_withdrawal_transactions()?;

        Ok(Self)
    }

>     pub fn build(self) -> Result<Bundle, OutgoingBundleError> {

Author: In the end, `validate` and `build` from `IncomingBundleBuilder ` and `SealedOutgoingBundleBuilder ` should be the same. What can we do for it ?

Reviewer 4: We can probably implement one in terms of the other. I have made a note in the sample code, but would leave this to the implementation phase.

        unimplemented!()
    }
}
```

## How this is used

> In this section, we describe the algorithms needed to build a `Bundle`. The work flow depends on whether one is

Author: > In this section, we describe the algorithms needed to build a `Bundle`.

Doesn't really make sense anymore.

Reviewer 4: Removed.
